### PR TITLE
fix for #914, rrdtool graph with --title segfaults

### PR DESCRIPTION
**Pull request by az143**
i've updated the graph_title_split() function a bit to not segfault and to handle various corner cases better (e.g. single title w/o delimiter didn't work before, back-to-back delimiters or leading or trailing ones).

### CODE REVIEW

File: src/rrd_graph.c [hunk 1]
```diff
@@ -6178,50 +6178,61 @@ image_title_t graph_title_split(
...
-    int delim = 1;
+    int count = 0;              /* line count */
+    int found_delim;
 
     retval.lines = malloc((MAX_IMAGE_TITLE_LINES + 1 ) * sizeof(char *));
-    length = strlen(title);
 
-    char *delims[6] = { "\n", "\\n", "<br>", "<br/>" };
-    printf("title: %s\n", title);
+    char *delims[4] = { "\n", "\\n", "<br>", "<br/>" };
```
Comment (oetiker) [inline src/rrd_graph.c]:
```C
char *delims[] = { "\n", "\\n", "<br>", "<br/>" };
```

would be more robust

File: src/rrd_graph.c [hunk 2]
```diff
@@ -6178,50 +6178,61 @@ image_title_t graph_title_split(
...
-        }
+       /*
+         search for next delimiter occurrence in the title,
+         if found, save the string before the delimiter for the next line & search the remainder
+
+         note: this logic cannot work if the title contains different delimiters,
+         as there's no guarantee that we'll find the earliest delimiter first!
+       */
+       found_delim = 0;
+       for(unsigned int i=0; i < 4; i++)
```
Comment (oetiker) [inline src/rrd_graph.c]:
```C
for(unsigned int i=0; i < sizeof(delims); i++)
```

**Comment by az143** [comment]
indeed, that's better - i had just opted for the minimum amount of changes that resolve the underlying issue.
anyway, both of your requested changes have been made.

**Comment by oetiker** [comment]
thanks!

**Comment by netniV** [comment]
As noted, this patch doesn't handle multiple delimiters and I verified that looking at the code.  

Whilst we would hope that a calling program will always use a single type, there are situations such as with Cacti, where multiple may be provided.  This can happen because the admin may entire a title style, but a device may also return a delimited in a value that gets used in a dynamic title.  

If the style picked by the device uses a delimiter earlier in the list, it would get broken up different and the string may include the operators intended break a line leading to unexpected output.  

Handling this corner case can be done with only a small improvement on what has been done here by I think:

1. Search for the position of each delimiter
2. Find the lowest index and substring from the current position to that location
3. Search again but just for the delimiter we have just used since we already have the lowest of the rest
4. Repeat part 2 until none found or the maximum lines is reached

If you feel it's worth it, Either I or @az143 (if they feel up to it) could apply that change.

**Comment by oetiker** [comment]
improvements are always welcome :) since we are supporting multiple line break chars, we might as well allow a mix ...

**Comment by netniV** [comment]
I'll take a look at later this week then unless @az143 feels he would rather do it.

**Comment by az143** [comment]
On Wed, 07 Nov 2018 15:33:41 -0800, Mark Brugnoli-Vinten writes:
>I'll take a look at later this week then unless @az143 feels he would
>rather do it.

please go for it; i'm a bit busy at the moment.

regards
az


-- 
Alexander Zangerl + GPG Key 2FCCF66BB963BD5F + http://snafu.priv.at/
"You are SUCH a useless use of a constant." -- flannelcat

**Comment by netniV** [comment]
@az143 and @oetiker, I have created a PR with what I believe should be working.  Feel free to review it and let me know.

**Pull request merged**